### PR TITLE
feat: partition nonce for miners who use multi-threads

### PR DESCRIPTION
**Pull request by yangby-cryptape**
Avoid same nonces for different threads.

### CODE REVIEW

File: miner/src/worker/mod.rs [hunk 1]
```diff
@@ -41,6 +43,17 @@ impl WorkerController {
     }
 }
 
+pub(crate) fn partition_nonce(id: u64, total: u64) -> Range<u64> {
+    let range = u64::max_value() / total;
```
Comment (doitian) [inline miner/src/worker/mod.rs]: name `span` is better

File: miner/src/worker/mod.rs [hunk 2]
```diff
@@ -41,6 +43,17 @@ impl WorkerController {
...
+    let start = range * id;
+    let end = match id {
+        x if x < total - 1 => range * (x + 1),
```
Comment (doitian) [inline miner/src/worker/mod.rs]:
```suggestion
        x if x < total - 1 => start + range,
```

File: miner/src/worker/mod.rs [hunk 3]
```diff
@@ -103,5 +118,11 @@ pub fn start_worker(
 }
 
 pub trait Worker {
-    fn run(&mut self, progress_bar: ProgressBar);
+    fn run(&mut self, nonce_range: Range<u64>, progress_bar: ProgressBar);
```
Comment (doitian) [inline miner/src/worker/mod.rs]: Alternative, we can just pass a random generator into it.

```suggestion
    fn run<G: FnMut() -> u64>(&mut self, rng: G, progress_bar: ProgressBar);
```

File: miner/src/worker/mod.rs [hunk 4]
```diff
@@ -41,6 +43,23 @@ impl WorkerController {
     }
 }
 
+fn partition_nonce(id: u64, total: u64) -> Range<u64> {
```
Comment (quake) [inline miner/src/worker/mod.rs]: integer division is round down, we can skip overflow checking:

```suggestion
fn partition_nonce(id: u64, total: u64) -> Range<u64> {
    debug_assert!(id < total);
    let span = u64::max_value() / total;
    let start = span * id;
    let end = start + span;
    (start..end)
}
```

Comment (yangby-cryptape) [reply miner/src/worker/mod.rs]: There still are few nonces in `end..u64::max_value()`.

Maybe the lucky nonce is one of them.

Comment (quake) [reply miner/src/worker/mod.rs]: Get, thanks.

btw, why set the parameter `id` and `total` type to u64? doesn't it should be usize?

Comment (yangby-cryptape) [reply miner/src/worker/mod.rs]: The output `nonce range` should be `u64`.

I think, both `u64` and `usize` are OK.
If I set `id` and `total` to `usize`, I can do cast at begin of the function `partition_nonce(..)`.

**Pull request merged**